<!-- Carousel
================================================== -->
<div id="myCarousel" class="carousel slide" data-ride="carousel">
  <!-- Indicators -->
  
<!-- /.carousel -->

<div class="container marketing">
  <!-- Three columns of text below the carousel -->
  <div class="row">
  <div class="container">
    <h1 class="text-center fit-h1" style="font-size: 4em; margin-bottom: 30px;">Gene Prioritization</h1>
  </div>
  </div>

  

  <!-- START THE FEATURETTES -->

  <hr class="featurette-divider">

  <div >
  <div >
  </div>
  <p class="lead">   Gene prioritizing schemes boost the power of identifying the most promising phenotype-associated based Machine Learning techniques among those clusters in normal conditions in different tissues (which each gene can have a normalized score of expression in Tissue expression database. Here we explain a new gene prioritization model based logistic regression application for candidate genes. </p>
  <div>
    
               <ul>
         <li> <p class="lead">Firstly, clustering methods are computing to segment our matrix-Expression-Similarity-based (Formula. 2) into k clusters (candidate’s genes whom contains only inputs and no desired output labels), in order each cluster represents a group of similar observations performed using the ward method and euclidean correlation for a given value of k as a possible (i.e., high intra-class similarity), whereas objects from different clusters are as dissimilar as possible (i.e., low inter-class similarity), in order to improve the initial partition. The algorithm can stop when the assignment of genes to clusters no longer changes or the specified maximum number of iterations has been reached. </p></li>

         <li><p class="lead">Then a classification as second step, which  algorithm contains both the inputs and the desired outputs (limited to a set of output fixed for each pathology), used to construct co-regulation , and link them into networks that prioritize candidate  genes as classification and to detect significant genetic interactions  as validation  form of the results in each expression trait attributable to differences in disease status,  cell or tissue type, ancestry, genetic background, experimental stimulus, or clinical variables. </p></li>

         
      
      </ul>
<ul><ul>
         <li> <p class="lead">For more details: All enriched terms are associated to their enrichment scores (e.g. p values), allowing the possibility to cross from high level concepts to  detailed  pathway  diagrams  showing  bio-molecular  events using groupGO(), enrichGO()  function. Genetic variation is also studied through our analysis, which the genome-wide association study (GWAS); based on GWAS Catalog used to tag variation across the genome and  enabling investigations to identify causal variants, and  variant-trait associations, which are mapped onto their chromosomal positions on the human genome. All traits are mapped into an ontology to improve searchability and establish targets for novel therapies.  </p></li>
         <li><p class="lead">Many of computational approaches has been developed to support the identification of the most promising candidates. oPOSSUM; web applications containing a great variety of the conserved non-coding regions of the promoters/enhancers were used to select the interaction of our candidate’s genes. In which, the interaction between genes and transcription factors (TFs) is important for understanding gene regulation and the origin of protein complexes components. </p></li>
      </ul>
      </ul>



<ul>
         <li> <p class="lead">Gene prioritization model could be formulated by arranging the candidate genes (i) in order of their tissues (t)  first, to be really sure to be associated with the pathology.  A typical gene prioritization application takes two inputs: a collection of evidence sources (i.e., databases of associations between genes, biological entities, and other biological/clinical entities as we fix) used to define a phenotype/trait of interest, and  enhancer/promoter interactions information,  that  extracted for a given genes and linked to each other with a normalized score (i.e.  the number of standard deviations by which the value of a raw score) that reflects “likelihood” of each gene to be responsible for the phenotype. Then a second factor which represent a prioritization mean factor (frequencies of occurrence of each entire).  The output of the algorithm is a list of candidate genes arranged according to calculated scores for each tissue. Every gene prioritization tool represents a unique combination of evidence sources, prioritization strategy and input requirements. Following the hypothesis that genes underlying similar tissues will share functional and phenotype characteristics, we incorporates logistic regression for any given  training genes which needed to be prioritized  (formula. 3).</p></li>

         

         
      
      </ul>



      <div class="container">
  <img src="images/formula33.jpg" alt="logo" class="center-block img-responsive"  style="margin-bottom: 80px;" />

<ul><ul>
         <li> <p class="lead">µ0  is the means of  expression-tissue-specific for a given gene with a fixed tissues. </p></li>
         <li><p class="lead">µ1,µ2,µ3,..µn represents  the means of  (frequencies of occurrence) for target genes for each process. </p></li>
         <li><p class="lead">X1, X2, X3,.. Xn represents the normalized expression values of biologival process, GWAS, and TFs, etc. </p></li>

      </ul>
      </ul>

<div class="container">
   <h4>Figure1: Data Warehouse Star Schema in SQL representing: A Fact Table representing prioritization model for different dimension tables. Note that, only PK are essential to start analysis in this application </h4>
  <img src="images/Capture111.jpg" alt="logo" class="center-block img-responsive"  style="margin-bottom: 80px;" />
 


</div>
    </div>
  </div>
 


  <!-- FOOTER -->
  <footer>
  <p class="pull-right"><a href="#">Back to top</a></p>
  
  </footer>

 </div>
  </div>
</div>
<!-- /END THE FEATURETTES -->
